datasets/dataset_plugin.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+from abc import ABC
+from typing import Callable, Dict, List, Optional
+
+from datasets.context import Context
+
+from .mode import Mode
+from .program_executor import ProgramExecutor
+
+
+class DatasetPlugin(ABC):
+    """
+    All dataset plugins derive from this class.
+    To register as a dataset they must decorate themselves with or call DatasetPlugin.register_plugin()
+    """
+
+    _executor: ProgramExecutor
+    # Context -> constructor_keys -> dataset plugin
+    _plugins: Dict[Context, Dict[set[str], DatasetPlugin]] = {}
+    _META_COLUMNS = ["run_id"]
+
+    def __init__(
+        self,
+        name: str,
+        logical_key: Optional[str] = None,
+        columns=None,
+        run_id: Optional[str] = None,
+        mode: Mode = Mode.Read,
+        attribute_name: Optional[str] = None,
+    ):
+        self.name = name
+        self.key = logical_key
+        self.mode = mode if isinstance(mode, Mode) else Mode[mode]
+        self.columns = columns
+        self.run_id = run_id
+        self._attribute_name = attribute_name
+        if not self._attribute_name:
+            self._attribute_name = name
+
+    @classmethod
+    def from_keys(cls, context: Optional[Context] = None, **kwargs) -> DatasetPlugin:
+        """
+        This is the factory method for datasets. Not directly used by the user.
+        For example usage please see test_from_keys()
+
+        :param context: If not specified it uses the current executor context.
+        :param kwargs: dataset constructor args
+        :return: found DatasetPlugin
+        """
+        dataset_args = set(kwargs.keys())
+        context_lookup: Context = context if context else cls._executor.context
+
+        default_plugin: Optional[DatasetPlugin] = None
+        max_intersect_count = 0
+        ret_plugin = None
+
+        for plugin_context in (
+            plugin_context for plugin_context in cls._plugins.keys() if context_lookup & plugin_context
```

Comment:
should this be  `if context_lookup == plugin_context`?

Reply:
A plugin may span multiple data contexts, `STREAMING | ONLINE`